File: src/go/rpk/pkg/cli/cmd/wasm/template/webpack.go
// Copyright 2020 Redpanda Data, Inc.
//
// Use of this software is governed by the Business Source License
// included in the file licenses/BSL.md
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0

package template

const webpack = `#!/usr/bin/env node
const webpack = require("webpack");
const fs = require("fs");

fs.readdir("./src/", (err, files) => {
  if (err) {
    console.log(err);
    return;
  }
  const webPackOptions = files.map((fileName) => {
    return {
      mode: "production",
      entry: { main: ` + "`./src/${fileName}`" + ` },
      output: {
        filename: fileName,
        libraryTarget: "commonjs2",
      },
      performance: {
        hints: false
      },
      target: 'node',
    };
  });
  webpack(webPackOptions, (err, stat) => {
    if (err) {
      console.log(err);
    }
    const info = stat.toJson();
    if (stat.hasErrors()) {
      console.error(info.errors);
    }
    if (stat.hasWarnings()) {
      console.warn(info.warnings);
    }
  });
});
`

func Webpack() string {
	return webpack
}

File: src/js/webpack.js
#!/usr/bin/env node
const webpack = require("webpack");
const os = require("os");
const fs = require("fs");
const path = require("path");
const nodeExternals = require("webpack-node-externals");

const webpackOptions = {
  context: __dirname,
  mode: "production",
  target: "node",
  entry: { main: `./modules/rpc/service.ts` },
  output: {
    path: path.resolve(__dirname, "dist"),
    filename: `main.js`,
    libraryTarget: "commonjs2",
  },
  resolve: {
    extensions: [".ts", ".js", ".node"],
  },
  module: {
    noParse: [/\/require-native.ts$/, /\/fast-crc32c\/loader\.js$/],
    rules: [
      {
        test: /\.ts?$/,
        use: "ts-loader",
        exclude: /node_modules/,
      },
      {
        test: /\.node$/,
        loader: "node-loader",
      },
    ],
  },
};
webpack([webpackOptions], (err, stat) => {
  if (err) {
    console.log(err);
  }
  const info = stat.toJson();
  if (stat.hasErrors()) {
    console.error(info.errors);
  }
  if (stat.hasWarnings()) {
    console.warn(info.warnings);
  }
});
